File: ios/App/App/VaultSseClient.swift
import Foundation

/// Native GLANCEvault SSE reader — the Swift sibling of Android's
/// `VaultSseClient` (android/.../sse/VaultSseClient.kt), ported from dayGLANCE's
/// production reader (dayglance-ios VaultSseBridge.swift). The transport,
/// lifecycle, and terminal-error policy are identical across the three readers —
/// only the renderer delivery differs (a message sink that [VaultSsePlugin]
/// wires to Capacitor's notifyListeners, not a WebView global).
///
/// WHY A NATIVE READER: the WKWebView cannot stream `text/event-stream` — vault
/// HTTP rides the buffered CapacitorHttp path (whole body as one string), and a
/// direct fetch from the capacitor:// app origin to the vault is CORS-blocked.
/// So the SHELL opens the stream: an authenticated `GET {vaultUrl}/events` with
/// the Bearer device token, read incrementally via `URLSession.bytes(for:)`,
/// framed into SSE event blocks, and each block pushed to the renderer. The
/// renderer reuses its EXISTING `parseSseFrame` + coalescer + drains
/// (src/sync/vaultEventStream.js) — this class owns ONLY the socket, its
/// lifecycle, and reconnect. A native HTTP client has no browser origin, so it
/// needs NO vault CORS change. POLLING in the renderer stays the correctness
/// backstop; these pushes only add instant drains on top.
///
/// LIFECYCLE — the reader runs only when BOTH are true: the renderer declared
/// SSE desired ([enable]/[disable]) AND the app is foreground ([setForeground],
/// wired from UIApplication notifications by the plugin). It drops on background
/// and reconnects with capped exponential backoff on any drop. Every transition
/// funnels through reconcile, which starts or cancels the SINGLE reader Task, so
/// there is never more than one connection. On each (re)connect the vault sends
/// its `ready` frame carrying the account's latest seq (`{"seq":N}`); it flows
/// through as a normal frame, so the coalescer reconciles anything missed while
/// disconnected.
final class VaultSseClient {

    /// Any-thread-safe message sink. All fields are strings:
    ///   {type:"open"} | {type:"frame", block} | {type:"closed"} |
    ///   {type:"error", message, code?}   (coded errors are TERMINAL)
    private let sink: ([String: String]) -> Void

    // ── State (guarded by `lock`, mirroring Android's @Synchronized) ─────────────
    private let lock = NSLock()
    // Framing-buffer cap (both the raw byte buffer and the decoded remainder). Real
    // vault events are tiny nudges; anything approaching 1 MB without a frame
    // delimiter is a misbehaving/hostile server, so we drop the connection rather
    // than grow unboundedly. Mirrors Android's SseFraming cap.
    private static let maxBufferBytes = 1_048_576
    private var desired = false
    // Foreground defaults true: the WebView only runs (and thus only calls start)
    // while the app is active, and the willEnterForeground notification won't fire
    // for the initial launch. Background transitions flip it false.
    private var foreground = true
    private var vaultUrl: String?
    private var token: String?
    private var accountId: String?
    private var readerTask: Task<Void, Never>?
    // Bumped on every (re)configure / stop so a stale reader that is mid-await
    // notices it is no longer the current generation and exits.
    private var generation = 0

    // Dedicated session: a request-idle timeout longer than the 20s server
    // heartbeat (a healthy idle stream never trips it, a silently-dead one is
    // detected within the window), and never a cached response. A stored `let`
    // (not lazy) so its first use from the reader Task can't race initialization.
    // waitsForConnectivity is OFF: our own reconnect + backoff (SseBackoff) owns
    // retries, matching Android's fail-and-reconnect model — so a drop surfaces
    // promptly instead of parking the request on the connectivity monitor.
    private let session: URLSession = {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 60
        cfg.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        cfg.waitsForConnectivity = false
        return URLSession(configuration: cfg)
    }()

    init(sink: @escaping ([String: String]) -> Void) {
        self.sink = sink
    }

    // MARK: - API (called from VaultSsePlugin, any thread)

    /// Renderer → "SSE desired ON", with the vault connection params.
    func enable(vaultUrl: String, token: String, accountId: String) {
        lock.lock()
        self.vaultUrl = trimTrailingSlashes(vaultUrl)
        self.token = token
        self.accountId = accountId
        self.desired = true
        reconcileLocked()
        lock.unlock()
    }

    /// Renderer → "SSE desired OFF".
    func disable() {
        lock.lock()
        desired = false
        reconcileLocked()
        lock.unlock()
    }

    /// App foreground/background (wired from UIApplication notifications).
    func setForeground(_ fg: Bool) {
        lock.lock()
        foreground = fg
        reconcileLocked()
        lock.unlock()
    }

    /// Full teardown — plugin deallocated. No leaked connection or task.
    func shutdown() {
        lock.lock()
        desired = false
        foreground = false
        reconcileLocked()
        lock.unlock()
        session.invalidateAndCancel()
    }

    // MARK: - Reconcile (start/stop the single reader) — call with `lock` held

    private func reconcileLocked() {
        let shouldRun = desired && foreground
            && vaultUrl != nil && token != nil && accountId != nil
        if shouldRun {
            guard readerTask == nil, let url = vaultUrl, let bearer = token, let account = accountId else { return }
            generation += 1
            let gen = generation
            readerTask = Task { [weak self] in
                await self?.runLoop(url: url, bearer: bearer, account: account, generation: gen)
                self?.clearReaderTask(ifGeneration: gen)
            }
        } else if let task = readerTask {
            // Bump the generation so the (possibly mid-await) reader exits, and cancel
            // to unblock the byte stream — URLSession's async iteration throws on Task
            // cancellation.
            generation += 1
            task.cancel()
            readerTask = nil
        }
    }

    // MARK: - Reader loop (reconnect + backoff)

    private func runLoop(url: String, bearer: String, account: String, generation gen: Int) async {
        var backoff = SseBackoff()
        while !Task.isCancelled && currentGeneration() == gen {
            let openedAt = Date()
            do {
                try await connectAndRead(url: url, bearer: bearer, account: account, generation: gen)
            } catch is CancellationError {
                return
            } catch let term as SseTerminalError {
                // TERMINAL (auth failure / insecure URL / server predates /events):
                // stop the reader entirely, no reconnect. Push a coded event so the
                // renderer surfaces it exactly once instead of every 30s. A later
                // enable() (user fixed the token/URL) spins up a fresh reader normally.
                if Task.isCancelled || currentGeneration() != gen { return }
                push(["type": "error", "code": term.code, "message": term.message])
                return
            } catch {
                if Task.isCancelled || currentGeneration() != gen { return }
                push(["type": "error", "message": error.localizedDescription])
            }
            if Task.isCancelled || currentGeneration() != gen { return }
            // The stream ended (server close / idle timeout / drop). Tell the
            // renderer, then reconnect with backoff — reset only if it had been
            // healthy long enough.
            push(["type": "closed"])
            backoff.onClosed(openDurationMs: Date().timeIntervalSince(openedAt) * 1000)
            try? await Task.sleep(nanoseconds: UInt64(backoff.nextDelayMs()) * 1_000_000)
        }
    }

    private func connectAndRead(url: String, bearer: String, account: String, generation gen: Int) async throws {
        guard var comps = URLComponents(string: "\(url)/events") else { throw URLError(.badURL) }
        comps.queryItems = [URLQueryItem(name: "accountId", value: account)]
        guard let target = comps.url else { throw URLError(.badURL) }

        // Belt-and-braces (the settings form is the primary gate): never send the
        // Bearer token over cleartext http on the public internet. https is always
        // fine; http is allowed only for loopback/LAN hosts. A refusal is TERMINAL —
        // reconnecting can't fix a bad scheme — so it takes the no-retry path above.
        guard Self.isSecureOrLanUrl(target) else {
            throw SseTerminalError(code: "insecure", message: "vault SSE refused: insecure http URL")
        }

        var req = URLRequest(url: target)
        req.httpMethod = "GET"
        req.setValue("Bearer \(bearer)", forHTTPHeaderField: "Authorization")
        req.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        req.timeoutInterval = 60

        let (bytes, response) = try await session.bytes(for: req)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        // 401/403 are auth failures — a revoked/invalid device token. Retrying can't
        // fix them (it would just reconnect forever at the 30s cap and re-hit the same
        // 401), so this is TERMINAL: stop the reader, push a distinct coded event so
        // the renderer surfaces it once. A fresh enable() (after the user fixes the
        // token) begins a new reader normally.
        if status == 401 || status == 403 {
            throw SseTerminalError(code: "auth", message: "vault SSE auth failed: \(status)")
        }
        // 404 means the server predates /events entirely (Express default route).
        // Same terminal logic: reconnecting cannot grow a route; the renderer's
        // polling covers the app identically. Mirrors the web transport's
        // SseUnsupportedError classification and Android's reader.
        if status == 404 {
            throw SseTerminalError(code: "unsupported", message: "vault SSE unavailable: server predates /events")
        }
        guard (200...299).contains(status) else {
            throw NSError(domain: "VaultSse", code: status,
                          userInfo: [NSLocalizedDescriptionKey: "vault SSE connect failed: \(status)"])
        }
        push(["type": "open"])

        // SSE framing at the BYTE level. We deliberately do NOT use bytes.lines
        // (AsyncLineSequence): it does not reliably yield the EMPTY lines that
        // delimit SSE event blocks, so a blank-line boundary is never seen and no
        // frame is ever emitted. Instead we accumulate the stream, normalize CRLF/CR
        // to LF, and split on the "\n\n" blank-line boundary ourselves — exactly like
        // Android's SseFraming and the renderer's drainSseBuffer. Decoding only at a
        // "\n" byte (0x0A, never part of a multibyte UTF-8 sequence) keeps UTF-8 safe
        // across chunk boundaries. Comment-only / heartbeat blocks (no data: line)
        // are dropped — matches Android's SseFraming.hasDataLine.
        var byteBuf = [UInt8]()
        var pending = ""
        for try await byte in bytes {
            if Task.isCancelled || currentGeneration() != gen { break }
            byteBuf.append(byte)
            // Cap the un-decoded byte buffer: a server that streams bytes but NEVER a
            // 0x0A ('\n') would otherwise grow this without bound (memory exhaustion).
            // Real events are tiny. Treat a breach as a stream failure (generic error →
            // backoff/reconnect), NOT terminal — a transient bad peer may recover.
            if byteBuf.count > Self.maxBufferBytes {
                throw SseStreamOverflowError()
            }
            guard byte == 0x0A else { continue }
            // String(decoding:as:) NEVER fails: invalid UTF-8 is replaced with U+FFFD
            // (matching Android's InputStreamReader), and byteBuf is ALWAYS cleared
            // here — a failable decode that left the buffer in place would stall the
            // stream permanently on a single bad byte.
            let chunk = String(decoding: byteBuf, as: UTF8.self)
            byteBuf.removeAll(keepingCapacity: true)
            pending += chunk.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\r", with: "\n")
            while let boundary = pending.range(of: "\n\n") {
                let block = String(pending[pending.startIndex..<boundary.lowerBound])
                pending = String(pending[boundary.upperBound...])
                if block.split(separator: "\n", omittingEmptySubsequences: false).contains(where: { $0.hasPrefix("data:") }) {
                    push(["type": "frame", "block": block])
                }
            }
            // Cap the framing remainder too: a server that sends '\n' bytes but never
            // the "\n\n" block delimiter would grow `pending` without bound. Same
            // treatment: fail the stream into backoff/reconnect.
            if pending.utf8.count > Self.maxBufferBytes {
                throw SseStreamOverflowError()
            }
        }
    }

    // MARK: - Push to the renderer

    private func push(_ msg: [String: String]) {
        // The sink (VaultSsePlugin → notifyListeners) is any-thread safe.
        sink(msg)
    }

    // MARK: - Helpers

    private func currentGeneration() -> Int {
        lock.lock(); defer { lock.unlock() }
        return generation
    }

    private func clearReaderTask(ifGeneration gen: Int) {
        lock.lock(); defer { lock.unlock() }
        // Only clear if we're still the current generation — a newer reconcile may
        // have already replaced the task.
        if generation == gen { readerTask = nil }
    }

    private func trimTrailingSlashes(_ s: String) -> String {
        var out = s
        while out.hasSuffix("/") { out.removeLast() }
        return out
    }

    // MARK: - URL transport-security allowlist (mirrors the renderer and Android reader)

    /// https is always allowed; http only for loopback/LAN hosts. Any other scheme is
    /// rejected. Keep in agreement with the renderer's policy and Android's reader.
    private static func isSecureOrLanUrl(_ url: URL) -> Bool {
        let scheme = (url.scheme ?? "").lowercased()
        if scheme == "https" { return true }
        if scheme != "http" { return false }
        return isLocalOrLanHost(url.host ?? "")
    }

    /// True for loopback / private-LAN / *.local hosts, for which cleartext http is
    /// acceptable. Foundation's URL.host already strips IPv6 brackets; the bracket
    /// strip below is defensive.
    private static func isLocalOrLanHost(_ rawHost: String) -> Bool {
        var host = rawHost.lowercased()
        if host.hasPrefix("[") && host.hasSuffix("]") { host = String(host.dropFirst().dropLast()) }
        if host.isEmpty { return false }
        if host == "localhost" || host.hasSuffix(".localhost") { return true }
        if host == "::1" { return true }
        if host.hasSuffix(".local") { return true }
        let parts = host.split(separator: ".", omittingEmptySubsequences: false).map(String.init)
        guard parts.count == 4 else { return false }
        let octets = parts.compactMap { Int($0) }
        guard octets.count == 4, octets.allSatisfy({ $0 >= 0 && $0 <= 255 }) else { return false }
        let a = octets[0], b = octets[1]
        if a == 127 { return true }              // 127.0.0.0/8 loopback
        if a == 10 { return true }               // 10.0.0.0/8
        if a == 192 && b == 168 { return true }  // 192.168.0.0/16
        if a == 172 && (16...31).contains(b) { return true } // 172.16.0.0/12
        return false
    }
}

/// A TERMINAL stream error: the reader must stop and NOT reconnect (retrying can't
/// help). `code` distinguishes the cause for the renderer ('auth' = revoked/invalid
/// token; 'insecure' = refused cleartext URL; 'unsupported' = server predates
/// /events). Mirrors Android's SseTerminalException.
private struct SseTerminalError: Error {
    let code: String
    let message: String
}

/// A NON-terminal stream failure: the SSE framing buffer exceeded its cap. Handled
/// like any read error — backoff + reconnect (a transient bad peer may recover).
private struct SseStreamOverflowError: Error {}

/// Capped exponential backoff with a stability reset — the Swift mirror of Android's
/// `SseBackoff` and the renderer's reconnect policy in `createVaultEventClient`.
///
/// A connection open at least `minStableMs` is treated as healthy and resets the
/// attempt counter, so a long-lived stream that finally drops reconnects promptly.
/// A connection that opens then drops immediately (a flap) keeps backing off, up to
/// `maxMs`, so a broken endpoint can never storm the vault with reconnects.
private struct SseBackoff {
    private let baseMs: Double
    private let maxMs: Double
    private let minStableMs: Double
    private var attempt = 0

    init(baseMs: Double = 1000, maxMs: Double = 30000, minStableMs: Double = 5000) {
        self.baseMs = baseMs
        self.maxMs = maxMs
        self.minStableMs = minStableMs
    }

    /// The delay before the NEXT reconnect, then advances the counter. `base * 2^attempt`,
    /// capped at `maxMs`; the shift is clamped so a large attempt count can't overflow.
    mutating func nextDelayMs() -> Double {
        let shift = min(max(attempt, 0), 30)
        let delay = min(maxMs, baseMs * Double(1 << shift))
        attempt += 1
        return delay
    }

    /// Reset the backoff only when the connection was open at least `minStableMs`
    /// (a healthy stream); a flap leaves the counter climbing.
    mutating func onClosed(openDurationMs: Double) {
        if openDurationMs >= minStableMs { attempt = 0 }
    }
}

File: android/app/src/main/java/com/lifeglance/app/sse/VaultSseClient.kt
package com.lifeglance.app.sse

import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import kotlin.coroutines.coroutineContext

/**
 * Native GLANCEvault SSE reader. Ported from dayGLANCE's production reader
 * (dayglance-android sse/VaultSseClient.kt); the transport, lifecycle, and
 * terminal-error policy are identical — only the renderer delivery differs
 * (Capacitor plugin events via [VaultSsePlugin], not a WebView global).
 *
 * The WebView cannot stream `text/event-stream` (vault HTTP rides the buffered
 * CapacitorHttp path), so the SHELL opens the stream instead: an authenticated
 * `GET {vaultUrl}/events` with the Bearer device token, read incrementally,
 * framed into SSE event blocks ([SseFraming]), and each block pushed into the
 * renderer via [frameSink] (which [VaultSsePlugin] wires to notifyListeners).
 * The renderer reuses its EXISTING `parseSseFrame` + coalescer + drains — this
 * class owns ONLY the socket, its lifecycle, and reconnect.
 *
 * Because this is a native HTTP client (no browser origin), it needs NO vault
 * CORS change. POLLING in the renderer stays the correctness backstop; these
 * pushes only add instant drains on top.
 *
 * LIFECYCLE — the reader runs only when BOTH are true:
 *   • the renderer has declared SSE desired ([enable] / [disable]), and
 *   • the Activity is foreground ([setForeground]).
 * It drops on background and reconnects with capped exponential backoff on any
 * drop ([SseBackoff]). Every transition funnels through [reconcile], which
 * starts or stops the SINGLE reader coroutine, so there is never more than one
 * connection.
 *
 * RECONNECT-RECONCILE — on each (re)connect the vault sends its `ready` frame
 * carrying the account's latest seq (`{"seq":N}`); it flows through [frameSink]
 * like any other, so the renderer's coalescer drains and catches anything missed
 * while disconnected.
 */
class VaultSseClient(
    private val frameSink: (String) -> Unit,
) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var readerJob: Job? = null

    // The connection currently being read, so cancellation (background / disable)
    // can force-close it and unblock the reader's blocking read().
    @Volatile private var activeConnection: HttpURLConnection? = null

    @Volatile private var desired = false
    @Volatile private var foreground = false
    @Volatile private var vaultUrl: String? = null
    @Volatile private var token: String? = null
    @Volatile private var accountId: String? = null

    /** Renderer → "SSE desired ON", with the vault connection params. */
    @Synchronized
    fun enable(url: String, bearer: String, account: String) {
        vaultUrl = url.trimEnd('/')
        token = bearer
        accountId = account
        desired = true
        reconcile()
    }

    /** Renderer → "SSE desired OFF". */
    @Synchronized
    fun disable() {
        desired = false
        reconcile()
    }

    /** Activity foreground/background (handleOnStart / handleOnStop). */
    @Synchronized
    fun setForeground(fg: Boolean) {
        foreground = fg
        reconcile()
    }

    /** Full teardown — Activity destroyed. No leaked connection or coroutine. */
    @Synchronized
    fun shutdown() {
        desired = false
        foreground = false
        stopReader()
        scope.cancel()
    }

    private fun shouldRun(): Boolean =
        desired && foreground && vaultUrl != null && token != null && accountId != null

    private fun reconcile() {
        if (shouldRun()) startReader() else stopReader()
    }

    private fun startReader() {
        if (readerJob?.isActive == true) return
        val url = vaultUrl ?: return
        val bearer = token ?: return
        val account = accountId ?: return
        readerJob = scope.launch { runLoop(url, bearer, account) }
    }

    private fun stopReader() {
        readerJob?.cancel()
        readerJob = null
        // Unblock a reader parked in a blocking read(): coroutine cancellation alone
        // won't interrupt java.io, so force-close the socket.
        try { activeConnection?.disconnect() } catch (_: Exception) {}
    }

    private suspend fun runLoop(url: String, bearer: String, account: String) {
        val backoff = SseBackoff()
        while (coroutineContext[Job]?.isActive == true) {
            val openedAt = System.currentTimeMillis()
            try {
                connectAndRead(url, bearer, account)
            } catch (e: SseTerminalException) {
                // TERMINAL (auth failure / insecure URL): retrying can't help, so stop
                // the reader entirely — no reconnect. Push a coded event so the renderer
                // surfaces it exactly once instead of every 30s. A later enable() (user
                // fixed the token/URL) launches a fresh reader normally.
                if (coroutineContext[Job]?.isActive != true) return
                Log.w(TAG, "SSE terminal (${e.code}): ${e.message}")
                push(JSONObject().put("type", "error").put("code", e.code).put("message", e.message ?: e.code))
                return
            } catch (e: Exception) {
                if (coroutineContext[Job]?.isActive != true) return // cancelled → done
                Log.w(TAG, "SSE read error: ${e.message}")
                push(JSONObject().put("type", "error").put("message", e.message ?: "sse error"))
            }
            if (coroutineContext[Job]?.isActive != true) return
            // The stream ended (server close / read timeout / drop). Tell the
            // renderer, then reconnect with backoff — resetting it only if the
            // connection had been healthy long enough.
            push(JSONObject().put("type", "closed"))
            backoff.onClosed(System.currentTimeMillis() - openedAt)
            delay(backoff.nextDelayMs())
        }
    }

    private suspend fun connectAndRead(url: String, bearer: String, account: String) {
        val target = URL("$url/events?accountId=" + URLEncoder.encode(account, "UTF-8"))
        // Belt-and-braces (the settings form is the primary gate): never send the
        // Bearer token over cleartext http on the public internet. https is always
        // fine; http is allowed only for loopback/LAN hosts. A refusal is TERMINAL —
        // reconnecting can't fix a bad scheme. (Android's manifest also blocks
        // cleartext to non-local hosts; this fails EARLY and with a clear reason.)
        if (!isSecureOrLanUrl(target)) {
            throw SseTerminalException("insecure", "vault SSE refused: insecure http URL")
        }
        val conn = (target.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            setRequestProperty("Authorization", "Bearer $bearer")
            setRequestProperty("Accept", "text/event-stream")
            connectTimeout = 20_000
            // Longer than the ~20 s server heartbeat so a healthy idle stream never
            // trips it, but a silently-dead connection is detected and reconnected
            // within the timeout instead of parking forever.
            readTimeout = 60_000
            instanceFollowRedirects = true
        }
        activeConnection = conn
        try {
            val code = conn.responseCode
            // 401/403 are auth failures — a revoked/invalid device token. Retrying just
            // reconnects forever at the 30s cap and re-hits the same 401, so this is
            // TERMINAL: stop the reader and push a distinct coded event the renderer
            // surfaces once. A fresh enable() (user fixed the token) reconnects normally.
            if (code == 401 || code == 403) {
                throw SseTerminalException("auth", "vault SSE auth failed: $code")
            }
            // 404 means the server predates /events entirely (Express default route).
            // Same terminal logic: reconnecting cannot grow a route; the renderer's
            // polling covers the app identically. Mirrors the web transport's
            // SseUnsupportedError classification.
            if (code == 404) {
                throw SseTerminalException("unsupported", "vault SSE unavailable: server predates /events")
            }
            if (code !in 200..299) throw RuntimeException("vault SSE connect failed: $code")
            push(JSONObject().put("type", "open"))

            val framing = SseFraming()
            val reader = BufferedReader(InputStreamReader(conn.inputStream, Charsets.UTF_8))
            val chunk = CharArray(2048)
            while (coroutineContext[Job]?.isActive == true) {
                val n = reader.read(chunk) // blocks until data, EOF (-1), or read timeout
                if (n == -1) break         // server closed the stream
                for (block in framing.append(String(chunk, 0, n))) {
                    push(JSONObject().put("type", "frame").put("block", block))
                }
            }
        } finally {
            activeConnection = null
            try { conn.disconnect() } catch (_: Exception) {}
        }
    }

    private fun push(msg: JSONObject) {
        // frameSink is any-thread safe: VaultSsePlugin routes it through
        // notifyListeners, which queues onto the Capacitor bridge.
        frameSink(msg.toString())
    }

    // ── URL transport-security allowlist ─────────────────────────────────────

    /** https is always allowed; http only for loopback/LAN hosts; anything else no. */
    private fun isSecureOrLanUrl(url: URL): Boolean {
        val scheme = (url.protocol ?: "").lowercase()
        if (scheme == "https") return true
        if (scheme != "http") return false
        return isLocalOrLanHost(url.host ?: "")
    }

    /** True for loopback / private-LAN / *.local hosts, where cleartext http is ok. */
    private fun isLocalOrLanHost(rawHost: String): Boolean {
        var host = rawHost.trim().lowercase()
        if (host.startsWith("[") && host.endsWith("]")) host = host.substring(1, host.length - 1)
        if (host.isEmpty()) return false
        if (host == "localhost" || host.endsWith(".localhost")) return true
        if (host == "::1") return true
        if (host.endsWith(".local")) return true
        val m = Regex("""^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$""").find(host) ?: return false
        val octets = m.groupValues.drop(1).map { it.toInt() }
        if (octets.any { it > 255 }) return false
        val (a, b) = octets
        return when {
            a == 127 -> true               // 127.0.0.0/8 loopback
            a == 10 -> true                // 10.0.0.0/8
            a == 192 && b == 168 -> true   // 192.168.0.0/16
            a == 172 && b in 16..31 -> true // 172.16.0.0/12
            else -> false
        }
    }

    companion object {
        private const val TAG = "VaultSseClient"
    }
}

/**
 * A TERMINAL SSE error: the reader must stop and NOT reconnect (retrying can't
 * help). [code] distinguishes the cause for the renderer ('auth' =
 * revoked/invalid token; 'insecure' = refused cleartext URL; 'unsupported' =
 * server predates /events).
 */
class SseTerminalException(val code: String, message: String) : Exception(message)
